shim/src/package_manager.rs

```diff
@@ -0,0 +1,86 @@
+use crate::paths::AbsolutePath;
+use anyhow::{anyhow, Result};
+use serde::Deserialize;
+use std::fs;
+use std::path::PathBuf;
+
+#[derive(Debug, Deserialize)]
+struct PnpmWorkspaces {
+    pub packages: Vec<PathBuf>,
+}
+
+#[derive(Debug, Deserialize)]
+struct PackageJsonWorkspaces {
+    pub workspaces: Vec<PathBuf>,
+}
+
+pub enum PackageManager {
+    #[allow(dead_code)]
+    Berry,
+    Npm,
+    Pnpm,
+    #[allow(dead_code)]
+    Pnpm6,
+    #[allow(dead_code)]
+    Yarn,
+}
+
+impl PackageManager {
+    /// Returns a list of globs for the package workspace.
+    /// NOTE: We return a `Vec<PathBuf>` instead of a `GlobSet` because we
+    /// may need to iterate through these globs and a `GlobSet` doesn't allow that.
+    ///
+    /// # Arguments
+    ///
+    /// * `root_path`:
+    ///
+    /// returns: Result<Vec<PathBuf, Global>, Error>
+    ///
+    /// # Examples
+    ///
+    /// ```
+    ///
+    /// ```
+    pub fn get_workspace_globs(&self, root_path: &AbsolutePath) -> Result<Vec<PathBuf>> {
+        match self {
+            PackageManager::Pnpm | PackageManager::Pnpm6 => {
+                let workspace_yaml = fs::read_to_string(root_path.join("pnpm-workspace.yaml"))?;
+                let workspaces: PnpmWorkspaces = serde_yaml::from_str(&workspace_yaml)?;
+                if workspaces.packages.is_empty() {
+                    Err(anyhow!("pnpm-workspace.yaml: no packages found. Turborepo requires pnpm workspaces and thus packages to be defined in the root pnpm-workspace.yaml"))
+                } else {
+                    Ok(workspaces.packages)
+                }
+            }
+            PackageManager::Berry | PackageManager::Npm | PackageManager::Yarn => {
+                let package_json_text = fs::read_to_string(root_path.join("package.json"))?;
+                let package_json: PackageJsonWorkspaces = serde_json::from_str(&package_json_text)?;
+
+                if package_json.workspaces.is_empty() {
+                    Err(anyhow!("pnpm-workspace.yaml: no packages found. Turborepo requires pnpm workspaces and thus packages to be defined in the root pnpm-workspace.yaml"))
```

Comment:
I think this error is incorrect

Reply:
I took it directly from the [Go version](https://github.com/vercel/turborepo/blob/main/cli/internal/packagemanager/pnpm.go#L38). Should I just not error here instead?

Reply:
We don't require `pnpm`, we do require _some_ package manager that supports workspaces.

Reply:
Eh, ok, I guess we can update it when we start making heavier use of package managers.